### PR TITLE
TYP: Use Self instead of class-bound TypeVar in pyi files (remove Y019)

### PR DESCRIPTION
**Pull request by topper-123**
Progress towards #51233. Next up will be to use `Self`  in the pandas code base.

xref #51309.

CC: @twoertwein

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -117,3 +117,4 @@ dependencies:
 
   - pip:
       - sphinx-toggleprompt
+      - typing_extensions; python_version<"3.11"
```
Comment (twoertwein) [inline environment.yml:120]: I think we don't need to add it, as it is only used during type checking (and typing_extensions is, I think, a dependency of mypy).

File: requirements-dev.txt [hunk 2]
```diff
@@ -86,4 +86,5 @@ feedparser
 pyyaml
 requests
 sphinx-toggleprompt
+typing_extensions; python_version<"3.11"
```
Comment (twoertwein) [inline requirements-dev.txt:89]: I don't think we need this

File: environment.yml (hunk 1 above)
Comment (topper-123) [reply environment.yml:120]: Yes, typing_extension is a dependency of mypy, I just checked for that. We do not want to be explicit, now that we're importing fro typing_extensions?

Comment (twoertwein) [reply environment.yml:120]: I think even the stubs packages from typeshed (for example https://pypi.org/project/types-psutil/) do not have typing_extensions as a dependency (neither does pandas-stubs).

If it is a new dependency, it probably should be added to the whatsnew (there might be other places where it needs to be added as well).

**Review by twoertwein — APPROVED** [review]
The `Self` changes look good to me. Whether `typing_extensions` should be a new dependency or not, should be judged by someone else.

**Comment by topper-123** [comment]
> The `Self` changes look good to me. Whether `typing_extensions` should be a new dependency or not, should be judged by someone else.

+1 from me, I'm also not sure what is normal.

I looked at `typeshed`, and the project itself has `typing_extensions` as a test requirement, se the [requirements-tests.txt](https://github.com/python/typeshed/blob/main/requirements-tests.txt). OTOH, individual type stubs don't have it as a requirement AFAIKS, e.g. `types-protobuf` uses `typing_extensions` but has no dependencies...

File: environment.yml (hunk 1 above)
Comment (simonjayhawkins) [reply environment.yml:120]: we have had discussions about this before and the outcome was that we do not use typing_extensions.

IIUC we don't need it with the latest mypy (1.0.0) when checking with python 3.11

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]

File: pandas/_typing.py [hunk 3]
```diff
@@ -83,8 +84,13 @@
     # Name "npt._ArrayLikeInt_co" is not defined  [name-defined]
     NumpySorter = Optional[npt._ArrayLikeInt_co]  # type: ignore[name-defined]
 
+    if sys.version_info >= (3, 11):
+        from typing import Self
+    else:
+        from typing_extensions import Self  # pyright: reportUnusedImport = false
```
Comment (simonjayhawkins) [inline pandas/_typing.py:90]: I guess you could make Self equivalent to Any here. IIRC we did something similar before to enable the latest typing features

**Comment by topper-123** [comment]
This PR was made as a bit of a follow-up to #51309, where @jbrockmendel had a suggestion to use `TypeGuard` for a more precise typing experience. `TypeGuard` is only available from python 3.10, so he added `type_extensions` to `_typing` module. #51309 hasn't actually been pulled in yet, but I didn't sense that using `TypeGuard`/`typing_extensions` as controversial.

I was not not aware of the decision to not use `typing_extensions` (which is totally on me, I had a break from pandas so have not been aware of some thing happening in the community). Would it be possible to reconsider the decision to not use `typing_extensions`?

IMO `typing_extensions` will be very useful to use in Pandas, because:
* `typings_extensions` is just a backport of functionality from newer python version to allow mypy to type check using older python versions,
* it is a dependency of `mypy` itself already (so is already installed, if mypy is installed), is developed by Guido & company (so has good support), so it doesn't seem controversial to me to use it, and
*  Pandas is a very large code base, so more/better typing is IMO will always be helpful to our project.
?

**Comment by simonjayhawkins** [comment]
> Would it be possible to reconsider the decision to not use `typing_extensions`?

When we had the discussion before I was arguing for including it. Now that we have pandas-stubs and typing in the codebase could be considered as a glorified linter there are probably less arguments for. Also, it has not always ended up been a blocker for using features supported in the latest mypy by implementing some trivial workarounds in some cases.

cc @TomAugspurger @jorisvandenbossche @jreback from previous discussions.

**Comment by simonjayhawkins** [comment]
The previous decision was for a run-time dependency.

As the import is inside a TYPE_CHECKING block this implementation only makes it a dev dependency (just like the use of npt)?

File: pandas/_typing.py (hunk 3 above)
Comment (simonjayhawkins) [reply pandas/_typing.py:90]: We type check using a single version of Python. There are capabilities in mypy to set the Python version to check against regardless of the version used. We don't do that.

So i'm not sure when this import is actually needed.

**Comment by simonjayhawkins** [comment]
scanning back over the historic typing_extensions discussions, the issue with having it as a dev-only dependency was the the need to wrap in TYPE_CHECKING and was also before typing_extensions was an required dependency of mypy in #27646

**Comment by simonjayhawkins** [comment]
> IMO `typing_extensions` will be very useful to use in Pandas, because:
> 
> * `typings_extensions` is just a backport of functionality from newer python version to allow mypy to type check using older python versions,
> * it is a dependency of `mypy` itself already (so is already installed, if mypy is installed), is developed by Guido & company (so has good support), so it doesn't seem controversial to me to use it, and
> * Pandas is a very large code base, so more/better typing is IMO always helpful to our project.

adding to this list...

* pandas-stubs is not restricted by the minimum Python version support. In is desirable to keep the pandas-stubs and pandas type annotations in sync where possible/appreciate.

**Comment by jorisvandenbossche** [comment]
From looking at the linked issues, https://github.com/pandas-dev/pandas/issues/37119 might be the most recent-ish complete discussion issue about this (+ https://github.com/pandas-dev/pandas/issues/34869 as a precursor of it). We can maybe have the actual discussion about adding as dependency or not there?

From skimming that issue, what is still unclear to me is what the exact consequences are? Is this only a dev dependency (only contributors have to have it installed in their environment) and not a required runtime dependency? Or would it only be useful if it's actually a runtime dependency? 
(the issue has various conflicting messages about this, some say it's possible, others not, or that it limits its usefulness, ..) 
While from this PR it seems it is only a dev dependency, since it's not added to pyproject.toml.

Can someone summarize this over at https://github.com/pandas-dev/pandas/issues/37119?

**Comment by topper-123** [comment]
I've answered in https://github.com/pandas-dev/pandas/issues/37119 regarding my experience using `typing_extensions` inside the `if TYPE_CHECKING` block.

Wrt. whether `typing_extensions` is/should be a run time dependency or not, I've made if NOT a run time dependency in this PR (only imported when `TYPE_CHECKING` is True).

If there's an argument for making this available at run time i'm open to arguments, but I just didn't see the  need myself.  and it makes sense to always limit run time dependencies when possible.

**Review by simonjayhawkins — APPROVED** [review]
Thanks @topper-123 

doesn't appear there are any objections from anyone this time round.

if part of the argument to use `typing_extensions` is that it is a required dependency of mypy, do we need to explicitly include it in `environment.yml`

requires merge main

lgtm pending green

**Comment by topper-123** [comment]
I've merged main.

**Comment by WillAyd** [comment]
Wow great work @topper-123

**Comment by jbrockmendel** [comment]
nice!

**Comment by jorisvandenbossche** [comment]
Can someone update https://github.com/pandas-dev/pandas/issues/37119 with the outcome?

**Pull request merged**